File: two-pointers/two-sum-sorted/two-sum-sorted.py
"""Two Pointers: Two Sum on a sorted array"""


def two_sum_sorted(nums: list[int], target: int) -> tuple[int, int]:
    left, right = 0, len(nums) - 1
    while left < right:
        total = nums[left] + nums[right]
        if total == target:
            return left, right
        if total < target:
            left += 1
        else:
            right -= 1
    return -1, -1


if __name__ == "__main__":
    nums = [1, 2, 3, 4, 6]
    target = 6

    result = two_sum_sorted(nums, target)
    print(f"Indices: {result[0]}, {result[1]}")

    # Every case in tests/two-sum-sorted.json, the same file the other languages read.
    # harness/python is on PYTHONPATH via .vscode/tasks.json and .vscode/launch.json.
    from dsatest import run_json_cases, run_tests

    run_json_cases("two-sum-sorted", two_sum_sorted)
    raise SystemExit(run_tests())

File: two-pointers/two-sum-sorted/two-sum-sorted.ts
// Two Pointers: Two Sum on a sorted array
import { pathToFileURL } from "node:url";

export function twoSumSorted(nums: number[], target: number): [number, number] {
  let left = 0;
  let right = nums.length - 1;

  while (left < right) {
    const sum = nums[left] + nums[right];
    if (sum === target) return [left, right];
    if (sum < target) left++;
    else right--;
  }

  return [-1, -1];
}

function main(): void {
  const nums = [1, 2, 3, 4, 6];
  const target = 6;

  const result = twoSumSorted(nums, target);
  console.log(`Indices: ${result[0]}, ${result[1]}`);
}

// Runs the demo only when this file is executed directly, so two-sum-sorted.test.ts
// can import the solution without the demo firing.
if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  main();
}
